packages/components/PrefixBadgePicker/Picker.tsx

```diff
@@ -0,0 +1,138 @@
+import type { ComponentProps, ReactNode } from "react";
+import { Children, cloneElement, isValidElement, useRef, useState } from "react";
+import classNames from "classnames";
+import { motion } from "framer-motion";
+
+import { getSpacingClassNamesFromProps } from "../utils";
+import { SpacingProps } from "../types";
+
+import styles from "./Picker.module.css";
+
+export function Picker<Values extends any>({
+  value,
+  onChange,
+  children,
+  backgroundColor,
+  ...props
+}: {
+  value: Values;
+  onChange: (value: Values) => void;
+  children: ReactNode;
+  backgroundColor?: string;
+  className?: string;
+  style?: React.CSSProperties;
+  title?: string;
+} & SpacingProps) {
+  const parsedProps = getSpacingClassNamesFromProps(props);
+  const previousId = useRef(null);
+  const transitioning = useRef(false);
+  const [isActive, setIsActive] = useState(false);
+  const [isOpen, setIsOpen] = useState(false);
+  const transition = {
+    duration: isOpen ? 0.16 : 0.2,
+    opacity: { duration: isOpen ? 0.1 : 0.04 },
+  };
+
+  return (
+    <motion.div
+      {...parsedProps}
+      data-open={isOpen}
+      className={classNames(styles.Picker, parsedProps.className)}
+      title={isOpen ? undefined : "Select a prefix badge"}
+      initial={isOpen ? "opened" : "closed"}
+      animate={isOpen ? "opened" : "closed"}
+      onMouseEnter={() => setIsActive(true)}
+      onMouseLeave={() => !isOpen && !transitioning.current && setIsActive(false)}
+      variants={{
+        opened: {
+          transition: {
+            staggerChildren: 0.02,
+          },
+        },
+        closed: {
+          transition: {
+            staggerChildren: 0.02,
+          },
+        },
+      }}
+    >
+      <motion.div
+        layout
+        transition={transition}
+        className={styles.PickerFill}
+        style={{
+          borderRadius: 32,
+          backgroundColor: isActive ? "var(--badge-background--active)" : "var(--badge-background)",
+        }}
+      />
+      {Children.toArray(children).map(child => {
+        if (!isValidElement(child)) {
+          throw Error("Picker children must be a valid React element.");
+        }
+
+        if (!child.props.id) {
+          const name = typeof child.type === "string" ? child.type : child.type.name;
+          throw Error(`Immediate Picker child "${name}" must pass an "id" prop.`);
+        }
+
+        const isToggle = child.props.id === "toggle";
+        const showToggle = value === undefined && isToggle;
+        const isChildActive = value === child.props.id || showToggle;
+        const wasChildActive = previousId.current === child.props.id;
```

Comment:
All of this state is gross and I want to clean it up in the near future. It is handling the active state for the toggle and other children to display styles properly.

Reply:
Agreed. To me, this feels really complicated to read and reason about compared to the UI before before animation was added to this component.

Do you think we'll be adding this kind of interaction in many places? If we do, I hope we abstract it away somehow. Would love to hear your thoughts about this somehow– how we can do this sort of thing in a maintainable way.

Reply:
I don't see it happening in many places so these spots _should_ be better isolated, but as the team and codebase grows I know that can easily get out of hand. I envision us getting our design system in place which will take all of this into consideration and keep the complexity wrapped in lower-level components that can communicate efficiently when used in the application.